### DISCUSSION TITLE
Changing default router for existing graph

### DISCUSSION THREAD

**Discussion by GeorgiyZhuravlev** [clientIO/joint#2768] [Q&A] [answered]
### Introduction

Hi guys!

We want to let our customers fine-tune graphs by selecting router and other related to links default paper settings (`defaultRouter`, `defaultConnector` etc). 
Both for existing (sometimes pretty complex, with lots of shapes and links already) and new (empty) graphs.

We made a small tool for that and for ourselves, to explore the options and pick the right ones (draft):
<img width="652" alt="image" src="https://github.com/user-attachments/assets/2762f27e-3745-4257-84b0-463f7297ae2c">

The idea is simple - settings from that UI are passed to the paper:
```
const {defaultRouter, defaultAnchor, defaultConnector, ...} = settings;
paper.options.defaultRouter = defaultRouter;
paper.options.defaultAnchor = defaultAnchor;
...
```

It turned out that just assigning new values to `paper.options.default*` is not enough for changes to take place, and what helps is updating all LinkViews after settings are changed:
```
    ...
    graph.getLinks().forEach(link => {
      if (clearVertices) {
        link.vertices([]);
      }
      paper.findViewByModel(link).update();
    });
```

With this, links are "re-routed," but it seems not in the same way they do when we draw them with that router the first time or when we move cells they connect.
Let me show it with animation:
![_router](https://github.com/user-attachments/assets/44476f64-10c2-4111-867e-64b530717fcd)

You see - when I move the source/target cell a little bit - the link route becomes perfect. But not right after applying.

Thus, our question is this:
What else, other than calling `paper.findViewByModel(link).update()` we could do to help router "re-route"?

Thank you!

### Steps to reproduce

given

### Restrictions & Constraints

no

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
You should use [`linkView.requestConnectionUpdate()`](https://docs.jointjs.com/api/dia/LinkView/#requestconnectionupdate).

**Reply by GeorgiyZhuravlev** [reply]
No, `requestConnectionUpdate()` does the same for us (like `.update()` did)...
Same as in the animation shown above.
Does the latest version have changes for routers or suggested function works differently in `4.*`?
(we are on `3.7.3` now)

**Reply by kumilingus** [reply]
It looks like you should configure the `manhattan` router to ignore the background cells. What is the configuration of the router?

**Reply by GeorgiyZhuravlev** [reply]
Settings for `manhattan` are just defaults.

```
  'manhattan': {
    name: 'manhattan',
    args: {
      padding: 10,
      step: 10,
      maximumLoops: 2000,
    }
  },
```

**Reply by kumilingus** [reply]
And what are the elements in the background? Are those graph cells or just an SVG appended to the paper?
You should tell the `manhattan` router to ignore those elements. For that use `excludeTypes` option.

**Reply by GeorgiyZhuravlev** [reply]
those are Pools 
`bpmn2.Pool` but we gave them some colors

```
export const Pool = shapes.bpmn2.Pool.define(
  'gluu.Pool',
  defaultInstanceProperties,
);
```

**Reply by kumilingus** [reply, marked as answer]
Then add `excludeTypes: ['gluu.Pool']` _manhattan_ option.

The _manhattan_ router most likely always failed and the fallback route was used. If you drag the element in the embedding mode, it is first unmbedded, then moved, and embedded again on drop. The result of the _manhattan_ could be different if the element has / has no parent.

**Reply by GeorgiyZhuravlev** [reply]
> Then add excludeTypes: ['gluu.Pool'] manhattan option.

It works! 🕺 🔥 ❤️ 
Thank you very much, Roman!
You saved our day**s** 😄
